OpenRA.Mods.Common/Traits/Render/WithDamageOverlay.cs

```diff
@@ -45,55 +55,111 @@ public class WithDamageOverlayInfo : TraitInfo, Requires<RenderSpritesInfo>, IRu
 				throw new YamlException("Specifying WithDamageOverlay.Offset requires the BodyOrientation trait on the actor.");
 		}
 	}
 
-	public class WithDamageOverlay : INotifyDamage, INotifyCreated
+	public class WithDamageOverlay : ConditionalTrait<WithDamageOverlayInfo>, INotifyDamage, ITick
 	{
 		readonly WithDamageOverlayInfo info;
 		readonly Animation anim;
 
-		bool isSmoking;
+		bool isPlayingAnimation;
+
+		int delay = -1;
 
 		public WithDamageOverlay(Actor self, WithDamageOverlayInfo info)
+			: base(info)
 		{
 			this.info = info;
 			anim = new Animation(self.World, info.Image);
 		}
 
-		void INotifyCreated.Created(Actor self)
+		protected override void Created(Actor self)
 		{
 			var rs = self.Trait<RenderSprites>();
 			var body = self.TraitOrDefault<BodyOrientation>();
 
 			WVec AnimationOffset() => body.LocalToWorld(info.Offset.Rotate(body.QuantizeOrientation(self.Orientation)));
-			rs.Add(new AnimationWithOffset(anim, info.Offset == WVec.Zero || body == null ? null : AnimationOffset, () => !isSmoking),
-				info.Palette, info.IsPlayerPalette);
+			rs.Add(new AnimationWithOffset(anim, info.Offset == WVec.Zero || body == null ? null : AnimationOffset, () => !isPlayingAnimation));
 		}
 
 		void INotifyDamage.Damaged(Actor self, AttackInfo e)
 		{
-			if (!info.DamageTypes.IsEmpty && !e.Damage.DamageTypes.Overlaps(info.DamageTypes))
+			if (IsTraitDisabled
+				|| e.DamageState < info.MinimumDamageState
+				|| e.DamageState > info.MaximumDamageState)
+			{
+				isPlayingAnimation = false;
+				return;
+			}
+
+			// Getting healed.
+			if (e.Damage.Value < 0)
+				return;
+
+			if (!isPlayingAnimation && delay <= -1)
+			{
+				delay = Util.RandomInRange(self.World.SharedRandom, info.InitialDelay);
+				if (delay <= 0)
+					StartAnimation(self.World.LocalRandom);
+			}
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (delay < 0)
 				return;
 
-			if (isSmoking) return;
-			if (e.Damage.Value < 0) return; /* getting healed */
-			if (e.DamageState < info.MinimumDamageState) return;
-			if (e.DamageState > info.MaximumDamageState) return;
+			// Actor DamgageState may have changed.
+			if (self.GetDamageState() < info.MinimumDamageState || self.GetDamageState() > info.MaximumDamageState)
+				delay = -1;
+			else if (--delay <= 0)
+				StartAnimation(self.World.LocalRandom);
+		}
+
+		protected override void TraitDisabled(Actor self)
+		{
+			isPlayingAnimation = false;
+		}
+
+		void StartAnimation(MersenneTwister random)
+		{
+			isPlayingAnimation = true;
+			delay = -1;
+			PlayAnimation(info.GetSequences(random).GetEnumerator());
+		}
 
-			isSmoking = true;
-			anim.PlayThen(info.IdleSequence,
-				() => anim.PlayThen(info.LoopSequence,
-					() => anim.PlayThen(info.EndSequence,
-						() => isSmoking = false)));
+		void PlayAnimation(IEnumerator<string> sequences)
+		{
+			if (sequences.MoveNext())
+				anim.PlayThen(sequences.Current, () => PlayAnimation(sequences));
+			else
+				isPlayingAnimation = false;
```

Comment:
Shouldn't `IEnumerator<string>` be disposed after it has been exhausted?

Reply:
Is it worth it? then you will need allocate IEnumerator<string> everytime something trigger overlay.

Reply:
we already allocate a new one every time. Tbh I was not aware that you need to dispose of them, if you do then this setup is not ideal as this code is not guaranteed to run